File: ControlLib/src/main/java/org/marsroboticsassociation/controllib/motion/TrajectoryCurveSegment.java
package org.marsroboticsassociation.controllib.motion;

public interface TrajectoryCurveSegment {

    double startTime();

    double endTime();

    default double duration() {
        return endTime() - startTime();
    }

    double valueAt(double time);

    double slopeAt(double time);

    double minValue();

    double maxValue();

    TrajectoryCurveSegment shiftedBy(double deltaTime);

    TrajectoryCurveSegment clippedTo(double newStartTime, double newEndTime);
}

File: ControlLib/src/main/kotlin/org/marsroboticsassociation/controllib/motion/TrajectoryCurveSegment.kt
package org.marsroboticsassociation.controllib.motion

interface TrajectoryCurveSegment {
    fun startTime(): Double

    fun endTime(): Double

    fun duration(): Double = endTime() - startTime()

    fun valueAt(time: Double): Double

    fun slopeAt(time: Double): Double

    fun minValue(): Double

    fun maxValue(): Double

    fun shiftedBy(deltaTime: Double): TrajectoryCurveSegment

    fun clippedTo(newStartTime: Double, newEndTime: Double): TrajectoryCurveSegment?
}
